sklearn/metrics/pairwise.py

```diff
@@ -585,6 +597,273 @@ def cosine_distances(X, Y=None):
+    Y_num = None
+
+    if not y_none:
+        Y_cat = Y[:, categorical_features]
+        Y_num = Y[:, np.logical_not(categorical_features)]
+        # This is to normalize the numeric values between 0 and 1.
+        Y_num = np.divide(Y_num, max_of_numeric, out=np.zeros_like(Y_num),
+                          where=max_of_numeric != 0)
+    else:
+        Y_cat = X_cat
+        Y_num = X_num
+
+    # This is to ignore FutureWarnings, as numpy will change
+    # how object arrays are compared in _gower_distance_row,
+    # as there is no other proposal to execute this efficiently
+    # without to avoid this warning.
+    warnings.filterwarnings("ignore",
+                            "numpy equal will not check object identity" +
+                            " in the future")
+    for i in range(n_rows):
+        j_start = i
+
+        # for non square results
+        if n_rows != y_n_rows or not y_none:
+            j_start = 0
+
+        Y_cat[j_start:n_rows, :]
+        Y_num[j_start:n_rows, :]
+        result = _gower_distance_row(X_cat[i, :], X_num[i, :],
+                                     Y_cat[j_start:n_rows, :],
+                                     Y_num[j_start:n_rows, :],
+                                     n_cols)
+        D[i, j_start:] = result
+        if y_none:
+            D[i:, j_start] = result
+
+    return D
+
+
+def detect_categorical_features(X, categorical_features):
+
+    """Identifies the categorical columns of an array.
+
+    Parameters
+    ----------
+    X : array-like, or pandas.DataFrame, shape (n_samples, n_features)
+
+    categorical_features : optional array-like, shape (n_features)
+        Indicates with True/False whether a column is a categorical attribute.
+
+        Alternatively, the categorical_features array can be represented only
+        with the numerical indexes of the categorical attribtes.
+
+        If the categorical_features array is None, they will be identified in
+        X as boolean values.
+
+    Returns
+    -------
+    categorical_features : ndarray, shape (n_features)
+    """
+
+    n_rows, n_cols = X.shape
+
+    if categorical_features is None:
+        categorical_features = np.zeros(n_cols, dtype=bool)
+        for col in range(n_cols):
+            # In numerical columns, None is converted to NaN,
+            # and the type of NaN is recognized as a number subtype
+            if not np.issubdtype(type(X[0, col]), np.number):
+                categorical_features[col] = True
+    else:
+        categorical_features = np.array(categorical_features)
+
+    if np.issubdtype(categorical_features.dtype, np.integer):
+        new_categorical_features = np.zeros(n_cols, dtype=bool)
+        new_categorical_features[categorical_features] = True
+        categorical_features = new_categorical_features
+
+    return categorical_features
+
+
+def _gower_distance_row(xi_cat, xi_num, xj_cat, xj_num, n_cols):
+    """Computes the gower distances between a set of rows of X and Y
```

Comment:
single row x and set of rows Y... but hard to see how that relates to `xi`, `xj`. use the same naming

Reply:
Hi Joel, thanks for this review. I still did not finish my changes, but will take the comments you just made into account.